localstack/services/events/target.py

```diff
@@ -0,0 +1,303 @@
+import json
+import logging
+import uuid
+from abc import ABC, abstractmethod
+
+from botocore.client import BaseClient
+
+from localstack.aws.api.events import (
+    Arn,
+    Target,
+)
+from localstack.aws.connect import connect_to
+from localstack.utils import collections
+from localstack.utils.aws.arns import (
+    extract_service_from_arn,
+    firehose_name,
+    sqs_queue_url_for_arn,
+)
+from localstack.utils.aws.client_types import ServicePrincipal
+from localstack.utils.strings import to_bytes
+from localstack.utils.time import now_utc
+
+LOG = logging.getLogger(__name__)
+
+
+class TargetSender(ABC):
+    def __init__(
+        self,
+        target: Target,
+        region: str,
+        account_id: str,
+        rule_arn: Arn,
+        service: str,
+    ):
+        self.target = target
+        self.region = region
+        self.account_id = account_id
+        self.rule_arn = rule_arn
+        self.service = service
+
+        self._validate_input(target)
+        self._client: BaseClient | None = None
+
+    @property
+    def arn(self):
+        return self.target["Arn"]
+
+    @property
+    def client(self):
+        """Lazy initialization of internal botoclient factory."""
+        if self._client is None:
+            self._client = self._initialize_client()
+        return self._client
+
+    @abstractmethod
+    def send_event(self):
+        pass
+
+    def _validate_input(self, target: Target):
+        """Provide a default implementation that does nothing if no specific validation is needed."""
+        # TODO add For Lambda and Amazon SNS resources, EventBridge relies on resource-based policies.
+        pass
+
+    def _initialize_client(self) -> BaseClient:
+        """Initializes internal botocore client.
+        If a role from a target is provided, the client will be initialized with the assumed role.
+        If no role is provided the client will be initialized with the account ID and region.
+        In both cases event bridge is requested as service principal"""
+        service_principal = ServicePrincipal.events
+        if role_arn := self.target.get("role_arn"):
+            # assumed role sessions expires after 6 hours in AWS, currently no expiration in LocalStack
+            client_factory = connect_to.with_assumed_role(
+                role_arn=role_arn, service_principal=service_principal, region_name=self.region
+            )
+        else:
+            client_factory = connect_to(aws_access_key_id=self.account_id, region_name=self.region)
+        client = client_factory.get_client(self.service)
+        client = client.request_metadata(
+            service_principal=service_principal, source_arn=self.rule_arn
+        )
+        return client
```

Comment:
This looks better, thanks! As discussed, it is necessary to set the `events` service principal to make sure all requests are recognized as coming from events :)